pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -526,6 +571,28 @@ void connectionOpened(final ClientCnx cnx) {
     }
 
     /**
+     * Clear the internal receiver queue and returns the message id of what was the 1st message in the queue that was
+     * not seen by the application
+     */
+    private MessageIdImpl clearReceiverQueue() {
+        List<Message> currentMessageQueue = new ArrayList<>(incomingMessages.size());
+        incomingMessages.drainTo(currentMessageQueue);
+        if (!currentMessageQueue.isEmpty()) {
+            MessageIdImpl nextMessageInQueue = (MessageIdImpl) currentMessageQueue.get(0).getMessageId();
+            MessageIdImpl previousMessage = new MessageIdImpl(nextMessageInQueue.getLedgerId(),
+                    nextMessageInQueue.getEntryId() - 1, nextMessageInQueue.getPartitionIndex());
+            return previousMessage;
+        } else if (lastDequeuedMessage != null) {
+            // If the queue was empty we need to restart from the message just after the last one that has been dequeued
+            // in the past
+            return lastDequeuedMessage;
```

Comment:
Not exactly, if the queue was already empty, then we can rely on the `lastDequeuedMessage`, but if the queue has messages, it means the application is actually getting those messages and then updating the `lastDequeuedMessage`, though not atomically.

By draining the queue, we can establish with 100% accuracy what would be the next message , the one that the application has surely not got yet. With that, we can just take 1 step back and get the last-dequeued message id.